Allow `unquote-splice` to accept any false value as empty

woodrush opened the pull request:
The original `unquote_splice` received a form stored in the variable `f_contents`, and produced the code
```clojure
`(list ~f_contents)
```
This PR modifies this to the following code, providing an additional `var = self.get_anon_var()`:
```clojure
`(do
   (setv ~var ~f_contents)
   (if* (= ~var None)
        []
        (list ~var)))
```
I have ran and passed the `tox` tests a Fedora Linux with Python versions 2.7, 3.3.6, 3.4.6, 3.5.3, 3.6.2, pypy-5.7.1, installed through pyenv. I also ran
```clojure
(defmacro splicetest []
  `[1 2 ~@[3 4] ~@None 5 6 ~@(when False 0) 7 8 ~@(when True None) 9 10 ~@(when True [None]) 11 12 ~None])
(splicetest)
```
through `hy2py`, and got
```python
import hy
from hy import HyInteger, HyList

def _hy_anon_fn_1():
    _hy_anon_var_1 = [3, 4]
    _hy_anon_var_2 = None
    _hy_anon_var_3 = None
    _hy_anon_var_4 = None
    _hy_anon_var_5 = [None]
    return HyList((((((((((((((((([] + [HyInteger(1)]) + [HyInteger(2)]) + ([] if (_hy_anon_var_1 == None) else list(_hy_anon_var_1))) + ([] if (_hy_anon_var_2 == None) else list(_hy_anon_var_2))) + [HyInteger(5)]) + [HyInteger(6)]) + ([] if (_hy_anon_var_3 == None) else list(_hy_anon_var_3))) + [HyInteger(7)]) + [HyInteger(8)]) + ([] if (_hy_anon_var_4 == None) else list(_hy_anon_var_4))) + [HyInteger(9)]) + [HyInteger(10)]) + ([] if (_hy_anon_var_5 == None) else list(_hy_anon_var_5))) + [HyInteger(11)]) + [HyInteger(12)]) + [None]))
hy.macros.macro('splicetest')(_hy_anon_fn_1)
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, None, 11, 12, None]
```
which is the expected result. (The last two `~`s are supposed to be evaluated to `None`.)

Edit: Closes #1335 .

woodrush commented:
Note that this code assures that the form is evaluated only once using `setv`, and uses `self.get_anon_var()` to assure that unwanted captures of variables don't occur.

woodrush commented:
An alternative way would be to change https://github.com/woodrush/hy/blob/f90cb39a61359c54089b2251324a11cb353b3b96/hy/compiler.py#L703 so that it appends nothing when it receives `None`. This would mean that we change
```clojure
(+ [] contents)
```
to
```clojure
(do
  (setv ret [])
  (for [form contents]
    (setv var form)
    (if (None? var) (continue) (ret.append var)))
  ret)
```
Now, both of these solutions don't change the fact that the `None`-ness of the form is evaluated after the form is compiled to Hy AST. Therefore, in both solutions we must construct this code using `HySymbol` and `HyExpression` constructs, so the alternative way of changing L703 would be almost the same as this PR. Also, the `contents` that is received here is what is created at https://github.com/woodrush/hy/blob/f90cb39a61359c54089b2251324a11cb353b3b96/hy/compiler.py#L711 , where I've changed in this PR (it is added at https://github.com/woodrush/hy/blob/f90cb39a61359c54089b2251324a11cb353b3b96/hy/compiler.py#L721). Changing L703 would also be a slight overhead for expressions that are not `unquote-splice`ed since it would add redundant checks of `quotedform?`.

Kodiologist commented:
Cool, thanks. Don't forget tests. Also, change your commit message and PR title to be more descriptive. It's hard to remember what #1335 is.

woodrush commented:
Thanks! I ran tests with `tox` for all the versions in `tox.ini`. Oh, but I forgot to add tests that splice `None`, I'll work on that.

> Also, change your commit message and PR title to be more descriptive.

I see. Got that.

woodrush commented:
Thanks for the comments. I added some tests for splicing `None`, squashed and rewrote the commit messages. I also reformatted the code where it was more than 80 characters long.

I'm writing another test which tests for `(unquote None)`.

woodrush commented:
Hmm... I think I've hit an issue-ish thing but I'm not that sure.
I'm trying to write a test that compares `(quasiquote (unquote None))` with `(quote None)`,

```clojure
; Test 1
(setv qq (quasiquote (a b c (unquote (+ 1 3))
                      (unquote None)
                      (unquote (when False 1)))))
(assert (= (len qq) 6))
(assert (= qq (quote (a b c 4 None None))))
```

but this fails since `(quasiquote (unquote None))` becomes `<class 'NoneType'>` while `(quote None)` becomes `<class 'hy.models.HySymbol'>`. The same thing happens with the original master branch as well.

So I tried casting `None`,

```clojure
; Test 2
(setv qq (quasiquote (a b c (unquote (+ 1 3))
                      (hy.models.HySymbol (unquote None))
                      (hy.models.HySymbol (unquote (when False 1))))))
(assert (= (len qq) 6))
(assert (= qq (quote (a b c 4 None None))))
```

but this fails since `HySymbol` needs to be fed with a `str`. Same for the original master branch.

I wondered why the original test works with the integers, so I looked inside the definitions of `HyInteger`, and searched the codebase for
overrides of `__eq__`, but it seems that there's no overloading of `=` going on. I'm not sure why `(= (quote 1) 1)` is `True`
although `(type (quote 1))` is `<class 'hy.models.HyInteger'>` and `(type 1)` is `<class 'int'>`.

I'm not sure how, and also a little unsure if I should write this test for `unquote` somehow right now... any suggestions?

gilch commented:
I'm trying to think of ways to avoid that extra anon var. If we simply used `or`, it would work about as well. So you could use ``` `(list (or ~fcontents []))```. This is not quite the same, since any falsy value would be allowed to stand for an empty list in a splice. But I think this is fine. Empty sequences are falsy, so should be fine to substitute the empty list, since that's what they'd generate in a `list` call anyway. In Common Lisp, `NIL` is the one false value, and also the empty list. So there is about as good a reason to accept `False` as there is to accept `None` as the empty list. 

It's a little bit strange at first to think that `0` (or any custom falsy type via `__bool__`) is now also allowed in a splice as an empty list. But I think this is both harmless and perfectly consistent.

gilch commented:
> I'm trying to write a test that compares `(quasiquote (unquote None))` with `(quote None)`,

These aren't supposed to be the same though. ``` `~x``` is just `x`, not `'x`.

In a typical Lisp, `NIL` is the empty list and also a symbol whose named value is itself. But Hy is based on Python, so the symbol `'None` is different from it's value `None`. The former is of type `HySymbol`, while the latter is of the singleton type `NoneType`--Python's `None` being the only instance of `NoneType`.

gilch commented:
I came up with another option to avoid the anon var. We can use exactly the same rules as `cons` by actually using a `cons`.
```
`(list (rest (cons Ellipsis ~f_contents)))
```
Where `Ellipsis` could be anything.

kirbyfan64 commented:
So...is there a reason we can't just do `(list ~(or f_contents []))`.

gilch commented:
@kirbyfan64, if I understand you right, you'd try
```
to_add = HyExpression([HySymbol("list"), f_contents or []])
```
Does this work?

woodrush commented:
Hmm, I tried that but that doesn't work. Although 
```clojure
`[1 2 ~@0] ;-> `[1 2],
`[1 2 ~@None] ;-> Becomes an error.
```
This worked:
```python
                    to_add = HyExpression([
                        HySymbol("list"),
                        HyExpression([HySymbol("or"), f_contents, HyList()])])
```

woodrush commented:
> These aren't supposed to be the same though. `~x is just x, not 'x.

I see. Well then I guess I'll just leave the test since it doesn't make that much sense as a practical expression.

> It's a little bit strange at first to think that 0 (or any custom falsy type via __bool__) is now also allowed in a splice as an empty list.

Since we have `lif`, I think we could create `lor` and put that in the core macros, where stuff like `defn` and `if` are. It would serve as part of a basis of what Hy considers as `nil`-ish.

woodrush commented:
Speaking of core macros, when I use the bootstrap macros that are in `bootstrap.hy` in where I wrote `to_add`, tests in pypy fail. I tried using `if` at first but failed, but `if*` didn't. `defn` also passed all versions except pypy but failed with it.

gilch commented on NEWS:
```diff
@@ -1,6 +1,8 @@
 
    [ Language Changes ]
+   * Pythonic falsey values (in means of __bool__) can now be spliced
+     by `unquote-splice`
```
I can't tell what this means.

A better wording would be "The `unquote-splice` or `~@` form now accepts any false value as empty."

woodrush replied:
I agree. Changed to that.

gilch commented:
What is the `tests/a_test_bin.py` file for? It seems like a duplicate of `tests/test_bin.py`.

woodrush commented:
> What is the tests/a_test_bin.py file for? It seems like a duplicate of tests/test_bin.py.

Sorry, that was an artifact for me renaming it to avoid `tox` to run `test_bin`. I dropped it out.

Kodiologist commented:
Looking good!

For legibility, would you please edit your new tests to use `` ` `` and `~@` instead of `quasiquote` and `unquote-splice`? I realize this will be inconsistent with other tests in the same functions; you change those too if you like.

Would you please update the documentation, namely the section on `unquote-splice` in `api.rst`?

woodrush commented:
Thanks! The test now uses `` \` `` and `~@`. I also added the docs on `unquote-splice`, any feedback would be appreciated.

woodrush commented:
Awesome!

gilch commented on docs/language/api.rst:
```diff
@@ -1604,9 +1604,12 @@ unquote-splice
-the quasiquoted form. ``unquote-splice`` is aliased to the ``~@`` symbol.
+the quasiquoted form. ``unquote-splice`` can also be used when the value
+evaluates to a false value such as ``None``, ``False``, and ``0``, in which
+the value is treated as an empty list, therefore splicing it to an empty value.
```
I'd reword this as 
> evaluates to a false value such as ``None``, ``False``, or ``0``, in which case the value is treated as an empty list and thus does not splice anything into the form.

woodrush replied:
I see, that sounds better.

gilch commented on docs/language/api.rst:
```diff
@@ -1617,6 +1620,15 @@ the quasiquoted form. ``unquote-splice`` is aliased to the ``~@`` symbol.
+Here, the last example evaluates to ``(u'+' 1L 2L)``, since the condition
+``(< (nth nums 0) 0)`` is ``False``, which makes this ``if`` expression
+evaluate to ``None``, since the ``if`` expression here does not have a
+second body. ``unquote-splice`` then evaluates this as an empty value,
```
I'd reword this to 
```
evaluate to ``None``, because the ``if`` expression here does not have an 
else clause.
```

gilch commented on docs/language/api.rst:
```diff
@@ -1604,9 +1604,12 @@ unquote-splice
+the quasiquoted form. ``unquote-splice`` can also be used when the value
+evaluates to a false value such as ``None``, ``False``, or ``0``, in which
+case the value is treated as an empty list and thus does not splice anything
+into the form. ``unquote-splice`` is aliased to the ``~@`` symbol.
```
I'd change the last "symbol" to "syntax". I know that's how it was written before, but the equivalent of `(unquote-splice foo)` is `~@foo` rather than `(~@ foo)` as the current wording seems to imply.

gilch commented on docs/language/api.rst:
```diff
@@ -1617,6 +1620,15 @@ the quasiquoted form. ``unquote-splice`` is aliased to the ``~@`` symbol.
     ;=> (u'+' 1L 2L 3L 4L)
 
+    `[1 2 ~@(if (< (nth nums 0) 0) nums)]
+    ;=> (u'+' 1L 2L)
```
Python 2 is considered legacy support now. We'll likely be removing support for Python 2 soon. #1217. All outputs in the documentation should be in the Python 3 form by then.

I'm not asking you to fix it everywhere in this file, just in the section you've added to to make it consistent. I'll even approve the pull without this particular fix, but we might as well do it now instead of making it worse.

woodrush commented:
@gilch, thanks for the review. All the changes are included now, including the Python 2 -> Python 3 notation for the section I edited.

gilch commented:
I think it's good to merge now. @Kodiologist, are you still OK with this version and commit structure?

Kodiologist commented:
Yes, all seems fine to me. I'll merge.
